app/src/main/java/com/mobileasone/dagger2workshop/di/SampleAppModule.kt

```diff
@@ -0,0 +1,32 @@
+package com.mobileasone.dagger2workshop.di
+
+import com.mobileasone.dagger2workshop.data.local.NotesLocalMemoryRepository
+import com.mobileasone.dagger2workshop.data.network.NotesServiceApi
+import com.mobileasone.dagger2workshop.data.network.NotesServiceApiImpl
+import com.mobileasone.dagger2workshop.domain.repositories.NotesRepository
+import dagger.Binds
+import dagger.Module
+import javax.inject.Singleton
+
+/**
+ * SampleAppModule provides the way to instantiate all dependencies needed in the application
+ * scope.
+ */
+@Module(subcomponents = [
+    NoteListFragmentComponent::class,
+    DetailNoteFragmentComponent::class,
+    AddNoteFragmentComponent::class,
+    OtherFragmentBComponent::class,
+    OtherFragmentCComponent::class
+])
+abstract class SampleAppModule {
+
+    @Binds
+    @Singleton
+    abstract fun notesServiceApi(service: NotesServiceApiImpl): NotesServiceApi
```

Comment:
When we return directly the instance in the arguments, we can use @Binds annotation. The method has to be abstract so Dagger can generate the implementation.
@Provides methods are instance methods and they need an instance of our module in order to be invoked. If our Module is abstract and contains @Binds methods, Dagger will not instantiate our module and instead directly use the Provider of our injected parameter (see [this link](https://proandroiddev.com/dagger-2-annotations-binds-contributesandroidinjector-a09e6a57758f)).
@Provides methods don't work on abstract classes so we cannot mix @Provides and @Binds annotations.